Touch controls editor

K1nop1c0 opened the pull request:
<!-- What is the motivation for the changes of this pull request? -->
Supports editing touch controls ingame. General features:
1. Edit everything that can be changed by editing touch_controls.json
2. Move & zoom buttons on screen directly
3. Automatically avoid selected button being overlapped
4. Preview button layout under different visibilities
5. Browse & search for buttons by label, command or position

Other edits:
Use `m_JoystickCount` to support multi joysticks on screen so `swap-action` can act correctly

Previews:
![Screenshot_2025-08-30-18-27-37-097_org ddnet clienu](https://github.com/user-attachments/assets/435014af-31a3-47b4-aa07-339f852b2152)
![Screenshot_2025-08-30-18-27-54-531_org ddnet clienu](https://github.com/user-attachments/assets/d32e5e09-8b9b-48fa-b62c-8a42104bdf31)
![Screenshot_2025-08-30-18-28-08-491_org ddnet clienu](https://github.com/user-attachments/assets/8936e8e8-382e-4a8b-8015-974301e7744d)
![Screenshot_2025-08-30-18-28-17-323_org ddnet clienu](https://github.com/user-attachments/assets/f65705d8-7d5c-49b2-bcc0-089eabbc0e45)


<!-- Note that builds and other checks will be run for your change. Don't feel intimidated by failures in some of the checks. If you can't resolve them yourself, experienced devs can also resolve them before merging your pull request. -->

## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

K1nop1c0 commented:
The general preview is done now, tested ingame found no crashes as for now, it needs further tests.
Preview:
![Uploading Screenshot_2025-03-29-21-12-43-005_org.ddnet.clienu.jpg…]()

K1nop1c0 commented:
The general preview is done, tested ingame found no crashes, but it needs further tests.
Preview:
![Screenshot_2025-03-29-21-12-43-005_org ddnet clienu](https://github.com/user-attachments/assets/c6643b89-4aa4-467a-b442-5d2299f54424)
![Screenshot_2025-03-29-21-12-48-240_org ddnet clienu](https://github.com/user-attachments/assets/714b0c30-39a3-4486-8449-ea165d8ec697)

Robyt3 commented:
Huge thanks for working on this! I'll try to give more detailed comments over the next days.

One thing to keep in mind is that all UI should generally also work on 5:4 resolutions. The ingame editor UI looks quite full so it probably wouldn't work on those aspect ratios very well. Would probably be better to split some of these functions into different tabs inside the editor UI.

Why is the diff so large though? Looks like you commited different line endings or something like that. Please revert that and squash your commits so the diff is readable.

K1nop1c0 commented:
General Features:
0.A visibility selector to toggle each visibilities. When the checkbox in ingame menu is selected, the selected visibility will be used to decide if a button is visible. It will be updated according the real visibility upon selecting the checkbox.
1.Hold a button to select it, or add a new button directly in ingame menu, the new button will be automatically selected.
2.To de-select a button, hold in empty space or using the De-Select button in menu.
3.After selecting a button, start sliding anywhere using one finger to move it, it will automatically avoid overlapping with any other buttons. You can also use two fingers to zoom it.
4.When a button is selected, the button editor will be rendered in menus.
5.There is 3 main setting lists. For the X,Y,W,H editbox, only integers allowed, its length limitation is 6. All leading zero will be removed.
6.When editing the X,Y,W,H, you can see the selected button changing their size as well, and if your inputs are currently illegal, it will adjust automatically.
7.When the shape is circle, it currently still uses x,y,w,h for size. I will update that later.
8.The button's visibility has 3 types: gray, +, -. Directly touch the visibility to change it. Gray visibility means it doesn't affect if the button will be visible. +visibility means button won't be visible if the visibility is off, - means won't be visible if the visibility is on.
9.The button's behavior could be changed using that listbox. Note that when switching between bind, bind toggle, predefined, settings are always cached.
10.The bind setting and the first command member of bind toggle uses the same command. Switching between them will keep the settings.
11.Save: saving the current settings to a button.
12.Cancel: reload all settings from current selected button, meaning that all settings are removed.
13.New Button: will create a button with behavior bind, without label and command, label type = plain, shape = rect, unit rect = {0,0,50000,50000}(Will automatically be moved to empty space), visibility will be the same as last selected button. If no button selected, all visibility will be gray. If a button is newly created, but it's not saved, will fail creating a new one before saving.
14.Delete Button: the button is just deleted, no matter if it's saved.
15.C&P Button: the same as New Button, including failing, but it copies all selected button's value.
16.De-Select Button: de-select a button and remove all cached settings.
17.Whenever trying to save a button, client will check if all inputs are legal.
Illegal settings: 1.Bind toggle with less than 2 commands
    2.Position overlapping with other visible buttons under current visibility.
    3.Width or Height less than 50000 or more than 500000.
    4.Button out of bound.
Redundant settings: 1.Button has extra-menu-x visibility, but there's no button to toggle it. Note that if there're several, the error will merge them together, as "Extra Menu 1, 2, 3".
    2.Visibility Demo Player is used with other visibility except extra menu.
Illegal settings can't kill the save progress, redundant settings can be saved after confirm.
18.Settings being made in editor will be considered as unsaved changes. If you directly slide the button only for a small distance(dx + dy < 10000), this will not be considered as a change, will be automatically discarded when changing selected button. But if you zoom, it will be considered as changes made.
19.At any time if function FindPositionXY can't find an empty slot for the button, the button's width and height will be set to 50000, then try again. If still no place, then client will open the menu.
20.FindPositionXY's time complexity is o(n^2 * logn), need optimization.
21.Unsaved change marked in editor means there's change made in cached settings, while the m_vTouchButton hasn't been truly edited, unsaved change marked below means there's changes made in m_vTouchButton that haven't been written to touch_controls.json.
22.The New Button will not create a button until you actually save it.

K1nop1c0 commented:
![Image_1743326978896](https://github.com/user-attachments/assets/fa1813fb-6692-475a-8981-d5e1c3df40c6)
![Image_1743326986369](https://github.com/user-attachments/assets/cd85f064-ea76-43a1-a851-4b0bb99872cf)
![Image_1743326980986](https://github.com/user-attachments/assets/b393dfc0-c22b-4866-b71b-a0dd6b970045)
Changed the editor layout. The edit to touch_controls.json is for preventing overlapping.
Also fixed some problems, now save & cancel button works correctly when there's nothing to save.
X, Y ,W, H will turn red now if the input is illegal.
Packed some codes to function so looks better.

K1nop1c0 commented:
![Image_1743333797723](https://github.com/user-attachments/assets/328e203b-e67c-4ad4-b260-9fbb12067e8a)
![Image_1743333795143](https://github.com/user-attachments/assets/c9db54e8-ba56-498e-98dd-12b0d6901f90)
It now look like this :)
Fix not saving when deselecting the checkbox.

K1nop1c0 commented:
Screenshot ingame with 5:4 Resolution :)
![Polish_20250330_231137657](https://github.com/user-attachments/assets/b8e63ab8-a66b-49ea-bf97-9262337efca7)
![Polish_20250330_230904697](https://github.com/user-attachments/assets/982088e5-8130-4b3a-8fd8-86b2894aab41)
![Polish_20250330_230953835](https://github.com/user-attachments/assets/d3d5a617-3e55-45c7-999e-41d686258843)
The bottom ones seem to be narrow, tomorrow i'll adjust them

Robyt3 commented on src/game/client/components/menus.h:
```diff
@@ -850,5 +860,66 @@ class CMenus : public CComponent
+
+	std::string m_Error;
+
+public:
```
I currently find it hard to understand how everything works together because it looks like both `CMenus` and `CTouchControls` store some part of the information and they need to access each other's functions and variables a lot.

Try to move all of the state and variables to `CTouchControls`. Make all variables private by default. Functions in `CMenus` that then only use functions or variables of `CTouchControls` should be moved to `CTouchControls` instead. Ideally the only functions and variables in `CMenus` should be those specifically for handling how the settings work in the menus, but not how the touch controls themselves work.

In particular, the touch controls should not access the menus with `GameClient()->m_Menus`. It should be a one-way relation of the menus accessing the touch controls, but the touch controls should not know anything about the menus component.

K1nop1c0 replied:
Most m_Menus access are for the popup, others are some settings stored in CMenus. All cached settings are in CMenus while the button for rendering the cached settings is in CTouchControls. I guess I should move all the cached settings to CTouchControls, should I move these Popups as well?

Robyt3 reviewed (changes requested):
I think splitting the editor into multiple tabs would be better. It's pretty full right now so it would be hard to add or change much in the future. Also, the overall layout is rather confusing because there isn't really a hierarchy of buttons to follow. I suggest putting a button bar on top of the entire touch controls editor rect with maybe these tabs:

- "General" tab:
   - "Edit touch controls"
   - "Setting"
   - "Edit visibilities"
- "Buttons" tab:
    - All the settings for buttons.

Some code comments below, I'll try to go through this PR incrementally.

K1nop1c0 commented on src/game/client/components/menus.h:
```diff
@@ -850,5 +860,66 @@ class CMenus : public CComponent
+	CTouchControls::CTouchButton *m_NewSelectedButton = nullptr;
+
+	// The biggest value's length is shorter than 7
+	CLineInputBuffered<7> m_InputX;
+	CLineInputBuffered<7> m_InputY;
+	CLineInputBuffered<7> m_InputW;
+	CLineInputBuffered<7> m_InputH;
```
How does CLineInputNumber work? I found that using them in editbox you can still input characters that aren't digits.

K1nop1c0 commented:
Now it looks like this.
![屏幕截图 2025-04-03 192808](https://github.com/user-attachments/assets/ce6c1842-2d6b-47c0-96a4-5798319dd993)
Fix all m_Menus issues, and all members are private now.

K1nop1c0 commented:
Currently when calling popups in CTouchControls, it will use m_Menus.SetActive(true);, which means the ingame menu will be rendered for a single frame(Or more, depending on when the popup is dealt by the system). Usually I can see a flash.

Robyt3 reviewed (changes requested):
Lots of UI/UX related comments below.

## File, Buttons, Settings tabs

![image](https://github.com/user-attachments/assets/cf0e9047-f19f-4d67-862a-f44b8509b92b)

- Should use the same layout and functions like the normal menu tabs. Start with the File tab on the left side, then add tabs on the right. Or split all three tabs with equal width but no space between them. Compare with the Game, Players, ... ingame tabs.
- The buttons don't react to mouse hovering yet. Should use the same functions as for the ingame tabs so the buttons change color when hovered and pressed.
- Rename "Button" tab to "Buttons" and "Setting" tab to "Settings".

## File tab

- The two direct touch input related settings from this page should also be moved to the new Settings tab, with a new sub-tab "Direct touch".

## Settings tab

![image](https://github.com/user-attachments/assets/0e252f8a-85fb-46e5-81d8-638b4fac1598)

- I would move the Visiblity and Color tabs from the ride side to the top. Also rename "Visibility" to "Preview Visibility" and "Color" to "Colors". These tab buttons should also behave like the other tab bars (react to hover and click).

### Visibility tab

- The list of visibilities currently scrolls really slowly with mouse wheel. The scrolling speed of scrollregions should generally be set to around 3 items are scrolled by moving the mouse wheel once.
- I'd remove the "Edit Visibilities" label when instead moving the tabs to the top so the tab title itself is descriptive enough.
- The items in the list don't respond the mouse hovering and clicking yet. The grey text of inactive visibilities is hard to read. I would instead use the established UI pattern of using a checkbox to indicate selection. The check boxes should be aligned on the left side then.
- I'd rather use a dark background instead of a white background, as most UI uses darker backgrounds for lists, e.g. list of skins.

### Direct touch

- Move the existing Direct touch input settings from the File tab here.

### Color tab

![image](https://github.com/user-attachments/assets/8ca51454-3403-49fa-b4a5-e7d4bfa5287e)

- "InActive" --> "Inactive".
- Decrease the font size or height of the color picker line UI elements or the reset button looks too large.
- I'd remove the "Edit Color Settings" label when instead moving the tabs to the top so the tab title itself is descriptive enough.
- "De-Select" --> "Deselect".
- When you edit a new button and then press the Deselect button it asks you to either Discard or Save your changes. But you should also be able to cancel instead, i.e. when you neither want to Discard nor Save your changes. But instead of adding a popup with three buttons it would be easier to only use buttons "No" and "Yes" and ask in the dialog whether the changes should be discarded or not. When pressing No nothing should be changed.
- Maybe not a bug but slightly weird: "Cancel" and "Delete button" do the same thing if are editing a new button.

## Buttons tab

![image](https://github.com/user-attachments/assets/3d68a29c-82af-4189-8835-5a55f77887ec)

- Increase the total width or move the buttons a bit so the Save and Cancel buttons are aligned on the left and right sides. Maybe split that row in 3 equal rects instead and place "Unsaved changes" in the middle so it's balanced.
- The "Position", "Visibility" and "Behavior" tabs should be moved to the top, above the content which they toggle. Should also use same tab functions so the handle mouse hover and click. These tab buttons also have the incorrect color when selected, only the selected tab should be darkened but for these tabs it's reversed.
- I'd rather use a dark background for the contents of the three sub-tabs instead of a white background, as most UI uses darker backgrounds for lists, e.g. list of skins.
- The backgrounds for the individual settings tabs should all have the same size and be aligned on the left and right side with the other UI elements.
- I think the touch buttons should be updated also when you save the changes and not only when you exit the editor. I got confused why my changes were not being shown after saving.
- When saving it asks about the demo player visibility. Maybe just make that the default selected visibility when creating new buttons?

### Position tab

- Align the Shape label and shape dropdown on the left and right side.
- Use "Rectangle" as the label in the dropdown menu.

### Visibility tab

![image](https://github.com/user-attachments/assets/4deb8085-4ea7-4b8f-b7ef-b12f34ca5718)

- Same as for the visiblity preview, the list scrolls too slowly with the mouse wheel and the grey text is barely readable. Since you need three states for every visibility it would be easier to a radio menu with three buttons (see `DoLine_RadioMenu`) for each visibility, e.g. "Ignore", "Include", "Exclude" instead of "grey text", "+visibility", and "-visibility".

### Behavior tab

![image](https://github.com/user-attachments/assets/44390055-06da-45f9-bbc7-b588597b1312)

- Using an icon label doesn't appear to work. I tried `\uf3ce` like in the Wiki, but it just shows `UF3CE` literally using the icon font.
- For predefined behavior "Extra Menu": Add a label for the number setting. If possible, use the icon font with minus and plus icons for the buttons.
- For bind toggle behavior: Instead of selecting the number, which is rather inconvenient, you could use a scrollregion and show all items at the same time. Then a button to add entries at the end. And buttons for every entry to delete that entry or move it up or down.

K1nop10 commented:
General previews:
![Image_1744884802501](https://github.com/user-attachments/assets/ad09b81d-a258-46c9-ac61-0c0dcf00af44)
![Image_1744884805830](https://github.com/user-attachments/assets/ce3dc7a7-bc89-46ef-95da-ffbc6ab69327)
![Image_1744884808758](https://github.com/user-attachments/assets/330219da-0cef-4df0-a36d-70faf81f5499)
![Image_1744884810755](https://github.com/user-attachments/assets/bec74820-23fb-42c2-8e2f-971529f08150)
![Image_1744884815068](https://github.com/user-attachments/assets/dd0dc5ef-5f6b-4d2b-816f-8b1829ea2773)
![Image_1744884817307](https://github.com/user-attachments/assets/bedee5a7-8797-4c9e-a0a2-2796cb301e5e)

K1nop10 commented:
Small fix to the errors(Rename EMenuType, prefixing members with MENU_) Fix style.
```diff

```

K1nop1c0 commented:
If the labels are good, I could add Localize() to them in next update.

K1nop1c0 commented:
Previews:
![Image_1745501654118](https://github.com/user-attachments/assets/55857010-3b5c-4d40-9d50-8cde9a13efc2)
![Image_1745501656150](https://github.com/user-attachments/assets/f8f40295-4eaf-4b6a-a6dd-215770ef1830)
![Image_1745501658992](https://github.com/user-attachments/assets/ff6e1349-f9e7-4786-8a12-feb4bb9e9fff)
![Image_1745501661156](https://github.com/user-attachments/assets/0de1d046-e945-47a3-b730-ff8ced8157bf)
![Image_1745501664847](https://github.com/user-attachments/assets/500384cd-de7a-41e7-a567-3f2e37dd8066)

K1nop1c0 commented:
New previews. Now the layout looks much better
![Image_1745568710331](https://github.com/user-attachments/assets/0ca2294c-129f-46cb-97e3-24a2faf28b64)
![Image_1745568711854](https://github.com/user-attachments/assets/65562595-9df6-43bc-a470-08dad49c2fbb)
![Image_1745568713489](https://github.com/user-attachments/assets/01125c2d-2702-4dd5-8305-82e322b16a01)
![Image_1745568714985](https://github.com/user-attachments/assets/f75854c2-e634-4715-8e6f-591121322fb8)

Robyt3 reviewed (changes requested):
## General

- Decrease margin at the bottom for all tabs so it's also consistent with left and right margins.

![image](https://github.com/user-attachments/assets/b0406de4-3772-4e4a-b660-9b229b9fe236)

## Buttons tab

![image](https://github.com/user-attachments/assets/dacf1d58-22fe-4680-aa6a-2fa887cab0c4)

- I would consider merging the sub-page that list all the buttons with this tab instead of having them separate:
    - Remove the "No button selected." label because that would already be implied because there are instructions to select a button.
    - "Select button" --> "Select button by touch"
    - Remove the "Preview buttons" button and put the contents of that page there directly (increase the height as necessary).

## Buttons tab (List of buttons)

![image](https://github.com/user-attachments/assets/e42bb171-867c-413a-a644-2708bf604bb5)

- "Exit preview" button: not necessary anymore when list merged into Buttons tab.
- I would remove the separate tabs for visible and invisible buttons and instead show them in one list with all visible buttons first and then all invisible buttons. Then add the label "Visible" or "Not visible" to every entry to show its visibility.
- I'd remove the "Preview label"/"Preview command" button and instead always show both, it looks like there should be enough space for this.
- The list is hard to use because the entry are large but mostly empty.
   - I think using a small font size and less spacing between the labels is fine for these entries. The touch buttons themselves also have a smaller font size.
   - If possible I'd try to put all information in just one line and add a header for the labels instead of duplicating them, so it works like the server browser list instead.
   - For the button command and labels use the UI label property to truncate with ellipsis at the end if the text doesn't fit.
   - You can still increase the width of the touch controls editor if necessary (as long as it still fits with a 5:4 resolution).
   - Replace the "Select with (arrow)" with only a "Select" button and move it to the right end of the row.

## Buttons tab (Button selected)

![image](https://github.com/user-attachments/assets/3f308bc4-dc2d-4d1c-bb25-928982ed0fd0)

- Remove the space between the Layout/Visiblity/Behavior tabs and the dark background entirely. Then change the corner rounding so the tabs and background fit together neatly like the global tabs with the global background. I think that should make it clearer which part of the UI is being controlled by the tabs. 
- Reduce the margin between buttons at the bottom and the dark background for the sub-tabs.
- "Unsaved Changes" --> "Unsaved changes"

### Layout/Visibility/Behavior tabs

- Also add/use consistent top/left/right margin inside of the dark background.

## Settings tab

![image](https://github.com/user-attachments/assets/13f1a0c2-08d9-4bc0-b366-897021e5d23f)

### Preview Visiblity

- This tab feels a bit out of place here. I would rename it "Preview" and move it up so we have "File, Buttons, Settings, Preview" as top-most tabs.
- The "Show all buttons" from the "Button Configs" tab also belongs in this tab instead. I would from top to bottom first put the explanation label "Preview button visiblity while the editor is active." then the "Show all buttons" checkbox and then the list of visibilities. Also, when "Show all buttons" is checked then the list should be disabled because I assume it has no effect then (see `SetActive` function for lists and scroll regions).

### Button Configs

- The tab title is unnecessary now, the settings only contain the color and direct touch settings for now.

K1nop1c0 commented:
The show all buttons function does nothing except showing all buttons, it will be automatically turned off when leaving that page, and also can't select a button while it is on.

K1nop1c0 commented:
Also I plan to create a draft PR to show how new behaviors or other things could be added by adding a simple behavior, so ppl can see the diff, also can be referenced in wiki. Is that good?

Robyt3 commented on src/game/client/components/touch_controls.h:
```diff
@@ -152,17 +159,43 @@ class CTouchControls : public CComponent
 
-	class CUnitRect
+	static constexpr const char *const SHAPE_NAMES[(int)EButtonShape::NUM_SHAPES] = {"rect", "circle"};
+	std::array<const char *, (int)EButtonShape::NUM_SHAPES> m_Shapes = {"Rectangle", "Circle"};
```
Same as `m_VisibilityStrings`.

Robyt3 commented on src/game/client/components/touch_controls.h:
```diff
@@ -481,13 +533,19 @@ class CTouchControls : public CComponent
 	ColorRGBA m_BackgroundColorInactive = ColorRGBA(0.0f, 0.0f, 0.0f, 0.25f);
 
+	// Default Inactive Color.
+	const ColorRGBA m_DefaultBackgroundColorInactive = ColorRGBA(0.0f, 0.0f, 0.0f, 0.25f);
```
Same as `m_VisibilityStrings`. This is a constant, no need for a member variable.

Robyt3 commented on src/game/client/components/touch_controls.h:
```diff
@@ -481,13 +533,19 @@ class CTouchControls : public CComponent
 	ColorRGBA m_BackgroundColorActive = ColorRGBA(0.2f, 0.2f, 0.2f, 0.25f);
 
+	// Default Active Color.
+	const ColorRGBA m_DefaultBackgroundColorActive = ColorRGBA(0.2f, 0.2f, 0.2f, 0.25f);
```
Same as `m_VisibilityStrings`. This is a constant, no need for a member variable.

Robyt3 commented on src/game/client/components/touch_controls.h:
```diff
@@ -559,6 +622,161 @@ class CTouchControls : public CComponent
+	CTouchButton *SelectedButton() const { return m_pSelectedButton; }
+	void SetSelectedButton(CTouchButton *TargetButton) { m_pSelectedButton = TargetButton; }
+	bool NoRealButtonSelected() const { return m_pSelectedButton == nullptr; }
+	void RemakeTmpButton() { m_pTmpButton = std::make_unique<CTouchButton>(this); }
+	CTouchButton *TmpButton() const { return m_pTmpButton.get(); }
```
`Tmp` is not very descriptive.

K1nop1c0 replied:
Renamed all tmp to sample, as the button is mainly used for previewing the editing button

Robyt3 commented on src/game/client/components/touch_controls.h:
```diff
@@ -559,6 +622,161 @@ class CTouchControls : public CComponent
+	public:
+		// Whether the issue is finished.
+		bool m_Finished = true;
+		CTouchButton *m_TargetButton = nullptr;
```
`m_pTargetButton`

Robyt3 commented:
> Also I plan to create a draft PR to show how new behaviors or other things could be added by adding a simple behavior, so ppl can see the diff, also can be referenced in wiki. Is that good?

You can do it if you want, but not really necessary. Showing it in the wiki might be helpful. Though it would have to be made clear that making additional input possible compared to the official client is not allowed.

K1nop1c0 commented on src/game/client/components/menus.h:
```diff
@@ -850,5 +851,165 @@ class CMenus : public CComponent
+
+	int m_EditBehaviorType = (int)EBehaviorType::BIND;
+	int m_PredefinedBehaviorType = (int)EPredefinedType::EXTRA_MENU;
+	int m_EditCommandNumber = 0;
```
Was used for choosing which command to edit in bind toggle, i'll delete that

K1nop1c0 commented on src/game/client/components/menus.h:
```diff
@@ -850,5 +851,165 @@ class CMenus : public CComponent
+	int m_PredefinedBehaviorType = (int)EPredefinedType::EXTRA_MENU;
+	int m_EditCommandNumber = 0;
+	int m_EditElement = 0; // 0 for shape&size, 1 for visibility, 2 for behavior.
+	int m_CachedNumber = 0;
```
Replaced with m_CachedExtraMenuNumber

Robyt3 commented on src/game/client/components/menus.h:
```diff
@@ -850,5 +851,165 @@ class CMenus : public CComponent
+	int m_EditCommandNumber = 0;
+	int m_EditElement = 0; // 0 for shape&size, 1 for visibility, 2 for behavior.
+	int m_CachedNumber = 0;
+	std::vector<std::unique_ptr<CLineInputBuffered<1024>>> m_vInputCommands;
+	std::vector<std::unique_ptr<CLineInputBuffered<1024>>> m_vInputLabels;
+	std::vector<CTouchControls::CBindToggleTouchButtonBehavior::CCommand> m_vCachedCommands;
```
Calling `Ui()->SetActiveItem(nullptr);` to deactivate the editbox would be better when an input is removed, otherwise it might lead to issues if the edit box is still selected without the input being active. The input will be deactivated automatically after the editbox is deactivated.

K1nop1c0 reviewed (commented):
There're several issues i found:
1. After editing visibility in preview tab, should we also update the current editing button's visibility so it would be visible with the previewing visibility? As checking if it is overlapping with other visible button uses the previewing visibility.
2. The FindPositionXY might need a stricter test, cause when I tested it using 400 buttons on screen, many buttons have the same X value and Y value. For example button1(1,1,1,1), button2(2,1,1,1), the two has the same Y value, which will reduce the calculation time by a lot. So note that when checking if the function is laggy, don't place buttons all in line. For example Button1(0,0,50000,50000), Button2(1,1,50000,50000)
3. Currently the hitbox of circle is still kinda weird. Maybe need a rework or somehow edit its CUnitRect.

K1nop1c0 commented:
Forget to remove the log_error in the code. You can see how much time the search function costs in the console by adding three numbers together.

K1nop1c0 reviewed (commented):
Haven't reproduced that on my side, the code uses pointers to store previewed buttons, so if there is nullptr it should crash before you select it. I'll try to give another test later.

Robyt3 reviewed (changes requested):
Mostly looks good to me in terms of functionality and UI. The code still needs some more review though. If anyone else also has code comments feel free to contribute as well.

Also needs some more testing to reproduce and fix the crash reported in https://github.com/ddnet/ddnet/pull/9965#pullrequestreview-2885491725.

Please avoid adding more features so we can hopefully review and merge this for next release.

Please also rebase on latest master and squash the changes into one commit.

K1nop1c0 commented:
New crash: New button, don't save then select another button by touch, choose save, then you'll see a ghost button that will crash if saving. This crash only happens when nothing else is done in the editor, or after you cancel/reset changes in file tab.
My PC is not on my side these days, will debug later

Robyt3 commented on src/game/client/components/menus.h:
```diff
@@ -20,9 +20,12 @@
 #include <game/client/component.h>
 #include <game/client/components/mapimages.h>
+#include <game/client/components/menus_ingame_touch_controls.h>
+#include <game/client/components/touch_controls.h>
```
The include of `touch_controls.h` should be unnecessary in this header file.

Robyt3 commented on src/game/client/components/menus_ingame.cpp:
```diff
@@ -237,7 +237,25 @@ void CMenus::RenderGame(CUIRect MainView)
+			{
+				m_MenusIngameTouchControls.m_pOldSelectedButton = GameClient()->m_TouchControls.SelectedButton();
+				m_MenusIngameTouchControls.m_pNewSelectedButton = nullptr;
+				PopupConfirm(Localize("Unsaved changes"), "Save all changes before turning off the editor?", Localize("Save"), "Cancel", &CMenus::PopupConfirm_TurnOffEditor);
```
`Localize` the message as well.

Robyt3 commented on src/game/client/components/menus_ingame_touch_controls.h:
```diff
@@ -0,0 +1,174 @@
+		CTouchControls::CBindToggleTouchButtonBehavior::CCommand m_CachedCommands;
+		CButtonContainer m_BindToggleAddButtons;
+		CButtonContainer m_BindToggleDeleteButtons;
+		CButtonContainer m_aLabelTypeRadios[3];
```
Is there a constant that can be used instead of `3`? If this is supposed to be the number of behaviors then `(int)EBehaviorType::NUM_BEHAVIORS`.

Robyt3 commented on src/game/client/components/menus_ingame_touch_controls.h:
```diff
@@ -0,0 +1,174 @@
+	// Used for creating ui elements.
+	std::array<CButtonContainer, (unsigned)CTouchControls::EButtonVisibility::NUM_VISIBILITIES> m_aVisibilityIds = {};
+	std::array<CButtonContainer, (unsigned)ESortType::NUM_SORTS> m_aSortHeaderIds = {};
+	std::array<CButtonContainer, 3> m_aEditElementIds = {};
```
`3` -> `(int)EElementType::NUM_ELEMENTS` ?

Robyt3 commented on src/game/client/components/menus_ingame_touch_controls.cpp:
```diff
@@ -0,0 +1,1499 @@
+	const char *pBehaviorType = pButton->m_pBehavior->GetBehaviorType();
+	if(str_comp(pBehaviorType, CTouchControls::CBindTouchButtonBehavior::BEHAVIOR_TYPE) == 0)
+	{
+		return static_cast<CTouchControls::CBindTouchButtonBehavior *>(pButton->m_pBehavior.get())->GetCommand();
```
Use `std::static_pointer_cast` for shared pointers instead of unboxing the pointer and then `static_cast`ing.

Example in editor code:

```
std::shared_ptr<CLayerQuads> pLayerQuads = std::static_pointer_cast<CLayerQuads>(pLayer);
```

Also in various other locations.

K1nop1c0 replied:
I'll check other places, but as for this location button uses std::unique_ptr for behaviors, using std::static_pointer_cast will make the original pointer invalid.

Robyt3 commented on src/game/client/components/touch_controls.h:
```diff
@@ -227,6 +241,8 @@ class CTouchControls : public CComponent
 		virtual void OnUpdate() {}
 		virtual void WriteToConfiguration(CJsonWriter *pWriter) = 0;
+		virtual const char *GetBehaviorType() const = 0;
+		virtual const char *GetPredefinedType() { return nullptr; }
```
This should not be a virtual function. It should only be declared by `CPredefinedTouchButtonBehavior`. Then, instead of calling this function directly, `std::static_pointer_cast` the behavior to `CPredefinedTouchButtonBehavior` after you checked `GetBehaviorType` like for the other behaviors.

Robyt3 commented on src/game/client/components/menus_ingame_touch_controls.cpp:
```diff
@@ -0,0 +1,1499 @@
+	}
+	// Used to decide if need to update the Samplebutton.
+	bool Changed = false;
+	CUIRect Left, A, B, C, EditBox, Block;
```
Avoid naming variables simply `A`, `B`, `C`. Try to give more descriptive names, e.g. `ButtonLayout`, `ButtonVisibility`, `ButtonBehavior` for the buttons here, then add more `CUIRect` variables as necessary below.

K1nop1c0 commented:
> New crash: New button, don't save then select another button by touch, choose save, then you'll see a ghost button that will crash if saving. This crash only happens when nothing else is done in the editor, or after you cancel/reset changes in file tab. My PC is not on my side these days, will debug later

This wild pointer crash bug has been fixed. Tested ingame, didn't find any weird behaviors.

K1nop1c0 commented:
<<fixed

K1nop1c0 commented:
Haven't checked if there're any bugs, my new PC dislikes ddnet and I wasted an entire day on this still can't enter the client.

K1nop1c0 commented:
The setter of `CLineInput` is const?

Robyt3 commented:
> The setter of `CLineInput` is const?

You mean `CLineInput::Deactivate`? Just make your function non-`const` if it doesn't work otherwise. I assume clang-tidy wanted the `const` there because the function doesn't technically modify `this` object (see also the failed CI run suggesting you add another `const` somewhere).

K1nop1c0 commented:
Replace `+=` with `append()`.

K1nop1c0 commented:
> Crash on `New button`.

Reason: Didn't change `emplace_back()` after changing `m_vBehaviorElemets` to `unique_ptrs`, so *nullptr happens. Fixed.

Change `RenderTouchButtonEditorWhileNothingSelected` to `RenderTouchButtonBrowser`.
Change `NewButton` in `CMenusIngameTouchControlsEditor` to `NewVirtualButton`, as this function does not create a real button.
Replace auto new button in `SaveCachedSettingsToTarget` when `pTargetButton` is nullptr by `dbg_assert` the `pTargetButton` must not be nullptr. Manually call `NewButton()` to create a button before saving settings. This also fix a bug that after new button then saving, selected button is still nullptr (not correctly set).

K1nop1c0 commented:
1. Remove useless time indicating comments in `CTouchControls::FindPositionXY()`
2. Add `nullptr` check for `malloc`
3. `CTouchControls::CTouchButton::UpdateScreenFromUnitRect()` now uses `CTouchControls::CalculateScreenFromUnitRect()` to avoid duplicate codes
4. `CTouchControls::DeleteButton()` -> `DeleteSelectedButton()`
5. `TPos` -> `BestPositionY`
6. Remove `FindExistingExtraMenus()` as it's not used anymore

Robyt3 commented on src/game/client/components/touch_controls.cpp:
```diff
@@ -1655,3 +1714,725 @@ void CTouchControls::WriteConfiguration(CJsonWriter *pWriter)
+			m_vOrder.erase(std::unique(m_vOrder.begin(), m_vOrder.end()), m_vOrder.end());
+			m_vZone.resize(m_vOrder.size() * 4, {-1, -1, 0, 0});
+			New(0, m_vOrder.size() - 2, 0);
+			m_pRes = (ivec2 *)malloc(m_vZone.size() * sizeof(ivec2));
+			dbg_assert(m_pRes != nullptr, "malloc for m_pRes failed.");
```
> Add `nullptr` check for `malloc`

We generally don't check the result of `malloc` except in special circumstances (map loading). I don't think you need to use `malloc` and `free` in this PR at all. You can use a `std::vector<ivec2> m_vRes` here instead.

Robyt3 commented on src/game/client/components/touch_controls.cpp:
```diff
@@ -1655,3 +1714,725 @@ void CTouchControls::WriteConfiguration(CJsonWriter *pWriter)
+		}
+		void InnerQuery(unsigned Start)
+		{
+			unsigned *pStack = (unsigned *)malloc(m_vZone.size() * sizeof(unsigned));
+			if(pStack == nullptr)
+			{
+				free(m_pRes);
+				dbg_assert(false, "malloc for pStack failed.");
+			}
```
Same, use a `std::vector<unsigned> vStack` instead of `malloc`.

K1nop1c0 commented:
Fix `Importing from clipboard` & `Discard changes` & `Reset to defaults` not updating color pickers
Again fix creating new button and saving by touch triggering dbg_assert

AssassinTee reviewed (commented):
This is a lot of code, but doesn't look too bad at a second glace. I'll do a deeper review soon

AssassinTee commented on src/game/client/components/touch_controls.h:
```diff
@@ -565,6 +591,121 @@ class CTouchControls : public CComponent
+	CUIRect CalculateScreenFromUnitRect(CUnitRect Unit, EButtonShape Shape = EButtonShape::RECT) const;
+
+	// Getters and setters.
+	bool UnsavedChanges() const { return m_UnsavedChanges; }
```
@ChillerDragon I'd add this then to the https://github.com/ddnet/ddnet/blob/master/CONTRIBUTING.md#getters-should-not-have-a-get-prefix

ChillerDragon replied:
I suppose. Will open a pr if another maintainer also likes the style.

AssassinTee commented:
I'd expect here a "No" option and just resume without saving any changes. You have to go back into the menu and deselect/drop changes yourself, handling this game in app is hard enough

<img width="2400" height="1080" alt="Screenshot_20250825-192504" src="https://github.com/user-attachments/assets/e317b858-a2d6-4874-b73d-c80e70f31fb6" />

I had a really hard time trying to understand this menu. I expected to be able to edit these values directly until I by accident double clicked on the name of one of them and another menu openend

<img width="2400" height="1080" alt="Screenshot_20250825-192336" src="https://github.com/user-attachments/assets/cee3604e-5942-47b6-a2c0-513ced89f6fd" />

K1nop1c0 commented:
> I'd expect here a "No" option and just resume without saving any changes. You have to go back into the menu and deselect/drop changes yourself, handling this game in app is hard enough. (This is the menu you get when you change a value but don't press "save" and try to resume without saving)
> 
> <img alt="Screenshot_20250825-192504" width="2000" height="1080" src="https://private-user-images.githubusercontent.com/821401/481736226-e317b858-a2d6-4874-b73d-c80e70f31fb6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYxNzg2OTIsIm5iZiI6MTc1NjE3ODM5MiwicGF0aCI6Ii84MjE0MDEvNDgxNzM2MjI2LWUzMTdiODU4LWEyZDYtNDg3NC1iNzNkLWM4MGU3MGYzMWZiNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyNlQwMzE5NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03NzQ4ODJhMGM5YjBhMTY4ZjJkMGNmN2NjMDRiZWM1MjMwZWUyNDNmZjIyZDViMzRhZDFkYjk4MTM2YWExMjk1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.rGLWcIsdEqqhtjJJTYKJ22yl4xv0_4XroYWfZjVla3A">
> I had a really hard time trying to understand this menu. I expected to be able to edit these values directly until I by accident double clicked on the name of one of them and another menu opened. If there is enough space a small E or Edit button would be nice I think
> 
> <img alt="Screenshot_20250825-192336" width="2000" height="1080" src="https://private-user-images.githubusercontent.com/821401/481736674-cee3604e-5942-47b6-a2c0-513ced89f6fd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYxNzg2OTIsIm5iZiI6MTc1NjE3ODM5MiwicGF0aCI6Ii84MjE0MDEvNDgxNzM2Njc0LWNlZTM2MDRlLTU5NDItNDdiNi1hMmMwLTUxM2NlZDg5ZjZmZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyNlQwMzE5NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNmMwYjc0OTgzYTEzMjc5ODNhY2YwYjkxZWJmOTVjNzBhMzgzNmZlNzUzOGU3MzI2MGQ4Y2Q3NDhhNDJkY2E5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.jdRM_DNBdwBepz6QhSpH5tFIjtabHM6FR3Oc663FlVw">
> Selecting a button by touch did not work for me (or I don't understand what it's supposed to do)

I suggest two options, the first is when a button box is selected, the button `Select button by touch` changes to `Edit button`. The second is changing the title `Long press on a touch button...` to `Double click to select a button`, and when selecting button by touch, render another message saying `Long press on a...`. How do you think?

K1nop1c0 commented:
CListBox is used in the browser, so double click on an element could select the button. To select a button by touch you need to long press on it

AssassinTee commented:
> We use Save+Discard when players are selecting&moving&resizing button by touch, as these behaviors are quite simple, players can easily decide saving&discarding changes. Even if they made a wrong decision, it will not be a big deal, just move it again.

I'd have expected a Cancel+Discard here, where canceling brings you back to the settings so you can save it again and Discard, well, discards. I felt "stuck in the menu" there :D A bit comparable to closing the client and you have changes in the editor: You can close anyway or go back and do whatever you must

K1nop1c0 commented:
Hmm, maybe you are right, but ig we need more opinions on this. Or why not add a popup with 3 buttons, 4 buttons, even DIY the popup page?

K1nop1c0 commented:
Kinda concerned about the misleading message there though

<img width="767" height="368" alt="image" src="https://github.com/user-attachments/assets/7e107219-eacc-44a9-ba33-393cd2e18a10" />

Should be double click in the browser, long press while selecting by touch

K1nop1c0 commented:
Please tell me if there're anything more need to be fixed. I'll be preparing & participating an algorithm contest ICPC soon, would probably have no free time from September 1. to 14.

Robyt3 commented on src/game/client/components/menus_ingame_touch_controls.cpp:
```diff
@@ -0,0 +1,1525 @@
+		{
+			str_format(aBuf, sizeof(aBuf), "%s:", Localize("Label type"));
+			Ui()->DoLabel(&LeftButton, aBuf, FONTSIZE, TEXTALIGN_ML);
+			int NewButtonLabelType = (int)m_vBehaviorElements[0]->m_CachedCommands.m_LabelType;
+			MiddleButton.VSplitLeft(MiddleButton.w / 3.0f, &LeftButton, &MiddleButton);
+			MiddleButton.VSplitMid(&MiddleButton, &RightButton);
+			if(GameClient()->m_Menus.DoButton_Menu(&m_vBehaviorElements[0]->m_aLabelTypeRadios[0], apLabelTypes[0], NewButtonLabelType == 0 ? 1 : 0, &LeftButton, BUTTONFLAG_LEFT, nullptr, IGraphics::CORNER_L))
+				NewButtonLabelType = 0;
+			if(GameClient()->m_Menus.DoButton_Menu(&m_vBehaviorElements[0]->m_aLabelTypeRadios[1], apLabelTypes[1], NewButtonLabelType == 1 ? 1 : 0, &MiddleButton, BUTTONFLAG_LEFT, nullptr, IGraphics::CORNER_NONE))
+				NewButtonLabelType = 1;
+			if(GameClient()->m_Menus.DoButton_Menu(&m_vBehaviorElements[0]->m_aLabelTypeRadios[2], apLabelTypes[2], NewButtonLabelType == 2 ? 1 : 0, &RightButton, BUTTONFLAG_LEFT, nullptr, IGraphics::CORNER_R))
+				NewButtonLabelType = 2;
+			if(NewButtonLabelType != (int)m_vBehaviorElements[0]->m_CachedCommands.m_LabelType)
```
Use enum `CButtonLabel::EType` instead of `int` and its literals instead of 0, 1, 2. Also in other places.

K1nop1c0 replied:
Changed two `NewButtonLabelType`s and array<int,...> `m_aCachedVisibilities` to array<enum>

Robyt3 commented on src/game/client/components/menus_ingame_touch_controls.cpp:
```diff
@@ -0,0 +1,1525 @@
+	s_VirtualVisibilityScrollRegion.Begin(&MainView, &ScrollOffset, &ScrollParam);
+	MainView.y += ScrollOffset.y;
+	std::array<bool, (size_t)CTouchControls::EButtonVisibility::NUM_VISIBILITIES> aVirtualVisibilities = GameClient()->m_TouchControls.VirtualVisibilities();
+	const char **apVisibilities = VisibilityNames();
```
`ppVisibilities`

Robyt3 commented on src/game/client/components/menus_ingame_touch_controls.cpp:
```diff
@@ -0,0 +1,1525 @@
+// Turn predefined behavior strings like "joystick-hook" into integers according to the enum.
+int CMenusIngameTouchControls::CalculatePredefinedType(const char *pType) const
+{
+	int IntegerType;
+	for(IntegerType = (int)EPredefinedType::EXTRA_MENU;
+		str_comp(pType, BEHAVIOR_FACTORIES_EDITOR[IntegerType].m_pId) != 0 && IntegerType < (int)EPredefinedType::NUM_PREDEFINEDS;
+		IntegerType++)
+		;
+	return IntegerType;
```
This looks rather unreadable. Separate string comparison into the loop body and break when finding the index instead of having an empty body.

K1nop1c0 commented:
GG, been waiting for this moment so long ;)
